graph/src/components/ethereum/network.rs

```diff
@@ -0,0 +1,273 @@
+use failure::{format_err, Error};
+use std::cmp::Ord;
+use std::collections::{BTreeMap, HashMap};
+use std::fmt;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use crate::components::ethereum::EthereumAdapter;
+pub use crate::impl_slog_value;
+
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum NetworkCapability {
+    Full,
+    Archive,
+    Traces,
+}
+
+impl FromStr for NetworkCapability {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "full" => NetworkCapability::Full,
+            "archive" => NetworkCapability::Archive,
+            "traces" => NetworkCapability::Traces,
+            s => return Err(format_err!("invalid network capability provided: {}", s)),
+        })
+    }
+}
+
+impl fmt::Display for NetworkCapability {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            NetworkCapability::Full => write!(f, "full"),
+            NetworkCapability::Archive => write!(f, "archive"),
+            NetworkCapability::Traces => write!(f, "traces"),
+        }
+    }
+}
+
+impl NetworkCapability {
+    pub fn sufficient_capability(
+        network: &Vec<NetworkCapability>,
+        required: &Vec<NetworkCapability>,
+    ) -> bool {
+        match required[..] {
+            [NetworkCapability::Full] => true,
+            [NetworkCapability::Traces] => network.contains(&NetworkCapability::Traces),
+            [NetworkCapability::Full, NetworkCapability::Traces] => {
+                network.contains(&NetworkCapability::Traces)
+            }
+            [NetworkCapability::Archive] => network.contains(&NetworkCapability::Archive),
+            [NetworkCapability::Archive, NetworkCapability::Traces] => {
+                network == &vec![NetworkCapability::Archive, NetworkCapability::Traces]
+            }
+            _ => false,
+        }
```

Comment:
`required` isn't guaranteed to be sorted, right? So we could be looking at any permutation of the capabilities. There must be a more elegant, less error-prone way of performing this check.

Since we're implementing `Ord`, `min`/`max` should help. _Something_ like

```rust
let max_network = network.iter().max();
let max_required = required.iter().max();
return max_network >= max_required;
```

Reply:
I'd suggest using a set for the capabilities, rather than a `Vec`. Or a bitfield :stuck_out_tongue:

Reply:
I was really hoping to do something like this, but I don't think it works because of the way there are two groups of capability types.  Let's try an example; what happens with a node that has [`Full`, `Trace`] and a subgraph with an `Archive` requirement.  If we order the enum: `Full`, `Archive`, `Trace` then the comparison will return true because `Trace` is higher then `Archive` but the node can't handle historic contract calls so this is incorrect.  

Now if we order the enum variants: `Full`, `Trace`, `Archive` the comparison will be incorrect in the case of node=`Archive` and subgraph requires [`Full`, `Trace`].

Reply:
@leoyvens a struct of `bool`s is a good idea. 🤔  
I was hoping to use a flexible data structure like a vec which we could add capabilities into, but really the..structure..of a struct will be useful here, as evidenced by the messiness of this comparison function. Realistically, new Ethereum node features will be very few and far between.

Reply:
@fordN You're right, there's not a single order between the capabilities. My bad.